src/dbnode/client/write_state.go

```diff
@@ -152,27 +170,53 @@ func (w *writeState) completionFn(result interface{}, err error) {
 		errStr := "missing shard %d in host %s"
 		wErr = xerrors.NewRetryableError(fmt.Errorf(errStr, w.op.ShardID(), hostID))
 	} else {
-		available := shardState == shard.Available
-		leaving := shardState == shard.Leaving
-		leavingAndShardsLeavingCountTowardsConsistency := leaving &&
-			w.shardsLeavingCountTowardsConsistency
-		// NB(bl): Only count writes to available shards towards success.
-		// NB(r): If shard is leaving and configured to allow writes to leaving
+		// in below conditions we consider the write success
+		// 1. writes to available shards towards success.
+		// 2. If shard is leaving and configured to allow writes to leaving
 		// shards to count towards consistency then allow that to count
-		// to success.
-		if !available && !leavingAndShardsLeavingCountTowardsConsistency {
-			var errStr string
-			switch shardState {
-			case shard.Initializing:
-				errStr = "shard %d in host %s is not available (initializing)"
-			case shard.Leaving:
-				errStr = "shard %d in host %s not available (leaving)"
-			default:
-				errStr = "shard %d in host %s not available (unknown state)"
+		// to success
+		// 3. If shardsLeavingAndInitializingCountTowardsConsistency flag is true then count the success on writing to both
+		// leaving and initializing as pair.
+		switch newCountTowardsConsistency(shardState,
+			w.shardsLeavingCountTowardsConsistency,
+			w.shardsLeavingAndInitializingCountTowardsConsistency) {
+		case availableCountTowardsConsistency:
+			w.success++
+		case shardLeavingIndividuallyCountTowardsConsistency:
+			w.success++
+		case shardLeavingAsPairCountTowardsConsistency:
+			// get the initializing host corresponding to the leaving host.
+			initializingHostID, ok := w.topoMap.LookupInitializingHostPair(hostID, w.op.ShardID())
+			if !ok || initializingHostID == "" {
```

Comment:
why is the `initializingHostID == ""` condition required? shouldn't the `ok` value returned be a sufficient check?